### PR TITLE
In tsc --b mode detect if root file of compilation is not root any more

### LINKED ISSUE #51866: Deleted source file not detected in composite mode
# Bug Report

### 🔎 Search Terms
composite deleted file tsconfig.tsbuildinfo

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about composite mode deleted files.
- Discovered in issue in current latest (4.9.4).

### 💻 Code

Can't use Playground or TypeScript Workbench since it appears to be a cache issue with tsconfig.tsbuildinfo.  Code must be built twice to see the issue.

Please clone https://github.com/drrnkin/typescript-composite-deleted-file.  It shows that the compiler fails to detect a deleted file, even though that deleted file should cause a compiler error.  This only happens in composite mode.

- In both the main and child folders, run "yarn" to install each.
- In the main folder, run "yarn build" to build both projects in composite mode.
- In the child folder, delete all "child2" files.
- In the main folder, run "yarn build" again and note that it doesn't detect the missing file "child2".

Note that compiling the child project directly (with "yarn tsc") detects the error, but it is missed in composite mode.

### 🙁 Actual behavior

Code that should not compile is not detected and reported.

### 🙂 Expected behavior

Should report an error.

### PR DESCRIPTION
**Pull request by sheetalkamat**
- Store the roots in BuildInfo. To make this compact, the roots is array that has elements as either [start, end] where fileIds from start to end are roots or single fileId. 
- In tsc -b detect if file was root before but not any more and trigger compilation based on that.
Fixes #51866

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot perf test

File: src/compiler/tsbuildPublic.ts [hunk 1]
```diff
@@ -1834,6 +1835,22 @@ function getUpToDateStatusWorker<T extends BuilderProgram>(state: SolutionBuilde
...
             newestInputFileTime = inputTime;
         }
+
+        if (buildInfoProgram) seenRoots.add(toPath(state, inputFile));
+    }
+
+    if (buildInfoProgram) {
+        if (!buildInfoVersionMap) buildInfoVersionMap = getBuildInfoFileVersionMap(buildInfoProgram, buildInfoPath!, host);
+        for (const existingRoot of buildInfoVersionMap.roots) {
+            if (!seenRoots.has(existingRoot)) {
```
Comment (jakebailey) [inline src/compiler/tsbuildPublic.ts:1845]: We don't do this for the opposite direction, i.e. if the old program had some file that the new one does not?

Comment (sheetalkamat) [reply src/compiler/tsbuildPublic.ts:1845]: We are checking if new program doesnt have a root that old program has
The check for new program having extra root happens when checking inputFile -> if it was already part of the old program here: https://github.com/microsoft/TypeScript/pull/52417/files#diff-0558af289c76f66611354da63a14152fe5cd3a72b3c3e59eb880808efc2df442L1824

**Comment by jakebailey** [comment]
I went and tested this PR using the linked issues repro (https://github.com/drrnkin/typescript-composite-deleted-file), and it didn't seem to work; I removed all of the child files and reran `yarn build` (modified to point to my local built folder), and none of the child files were recreated.

**Comment by sheetalkamat** [comment]
Here is the result from running on the repro:

```
c:\temp\typescript-composite-deleted-file\main>node C:\TypeScript\built\local\tsc.js -b -v
[1:45:22 PM] Projects in this build:
    * ../child/tsconfig.json
    * tsconfig.json

[1:45:22 PM] Project '../child/tsconfig.json' is out of date because output file '../child/tsconfig.tsbuildinfo' does not exist
[1:45:22 PM] Building project 'c:/temp/typescript-composite-deleted-file/child/tsconfig.json'...

[1:45:25 PM] Project 'tsconfig.json' is out of date because output file 'tsconfig.tsbuildinfo' does not exist

[1:45:25 PM] Building project 'c:/temp/typescript-composite-deleted-file/main/tsconfig.json'...


c:\temp\typescript-composite-deleted-file\main>del ..\child\child2.*

c:\temp\typescript-composite-deleted-file\main>dir ..\child\child2*
 Volume in drive C is Local Disk
 Volume Serial Number is 9EF5-3A92

 Directory of c:\temp\typescript-composite-deleted-file\child

File Not Found

c:\temp\typescript-composite-deleted-file\main>node C:\TypeScript\built\local\tsc.js -b -v
[1:45:51 PM] Projects in this build:
    * ../child/tsconfig.json
    * tsconfig.json

[1:45:51 PM] Project '../child/tsconfig.json' is out of date because buildinfo file '../child/tsconfig.tsbuildinfo' indicates that file '../child/child2.ts' was root file of compilation but not any more.

[1:45:51 PM] Building project 'c:/temp/typescript-composite-deleted-file/child/tsconfig.json'...

../child/child.ts:1:24 - error TS2792: Cannot find module '../child/child2.js'. Did you mean to set the 'moduleResolution' option to 'nodenext', or to add aliases to the 'paths' option?

1 import { child2 } from "../child/child2.js";
                         ~~~~~~~~~~~~~~~~~~~~

[1:45:52 PM] Project 'tsconfig.json' can't be built because its dependency '../child' has errors

[1:45:52 PM] Skipping build of project 'c:/temp/typescript-composite-deleted-file/main/tsconfig.json' because its dependency 'c:/temp/typescript-composite-deleted-file/child' has errors


Found 1 error.


c:\temp\typescript-composite-deleted-file\main>
```

**Comment by jakebailey** [comment]
Ah, I didn't delete the input files, I only deleted the output files. When delete _everything_, I get the error. I didn't realize the issue said to literally delete everything and that an error should occur. So, I guess it does work, oops.

I am however confused as to how deleting all of the output files (`rm child*.{js,d.ts}`) doesn't trigger a rebuild, but maybe that's another issue.

**Comment by sheetalkamat** [comment]
Thats a long standing design where we dont check all output file timestamps if writing tsbuildInfo for perf reasons. If you are deleting some or all outputs, delete buildInfo and run the build has been our guidance.

File: tests/baselines/reference/tsbuild/amdModulesWithOut/prepend-reports-deprecation-error.js [hunk 2]
```diff
@@ -447,6 +447,20 @@ declare const globalConst = 10;
...
+        [
+          2,
+          5
+        ],
+        [
+          "./file0.ts",
+          "./file1.ts",
+          "./file2.ts",
+          "./global.ts"
+        ]
```
Comment (jakebailey) [inline tests/baselines/reference/tsbuild/amdModulesWithOut/prepend-reports-deprecation-error.js:461]: My understanding is that this encoding is either a range of indexes over some array, or if noncontiguous, the filename from that array. Would it be more efficient to instead store either ranges or indexes, like:

```ts
[[2, 5], 7, 9]
```

Or, am I misinterpreting the format?

Comment (sheetalkamat) [reply tests/baselines/reference/tsbuild/amdModulesWithOut/prepend-reports-deprecation-error.js:461]: It is stored as ranges or indices. 

So here are examples: 
 `[[2,5],7,9]` => `2,3,4,5,7,9` eg https://github.com/microsoft/TypeScript/pull/52417/files#diff-5fbf0d47ec6990dcf27581b225204e9671350ee831847ee68d498fddfd852e7dR983
https://github.com/microsoft/TypeScript/pull/52417/files#diff-97195a8f223f06c8bc85c9b5faef6126b946d0a6dc3a47cf3bde7f80e099a444R507

 `[2,3,[7,9]]` => `2,3,7,8,9` eg: https://github.com/microsoft/TypeScript/pull/52417/files#diff-5fbf0d47ec6990dcf27581b225204e9671350ee831847ee68d498fddfd852e7dR603

`[2,[7,9]]` => `2,7,8,9` eg: https://github.com/microsoft/TypeScript/pull/52417/files#diff-5b7eeb5a2a5682fb3f3ddde640382cd676760cd49d3a0513bd1df1fd9c017389R115


And will add some more as a next commit to show these cases clearly

Comment (sheetalkamat) [reply tests/baselines/reference/tsbuild/amdModulesWithOut/prepend-reports-deprecation-error.js:461]: [c6d40cb](https://github.com/microsoft/TypeScript/pull/52417/commits/c6d40cb7d62342ada63c4776bec2aad4a997f6b5)

Comment (jakebailey) [reply tests/baselines/reference/tsbuild/amdModulesWithOut/prepend-reports-deprecation-error.js:461]: Oh, I think my confusion comes from the fact that I was expecting these to not encode filenames at all (to save space), but I guess that's not what's happening.

Comment (sheetalkamat) [reply tests/baselines/reference/tsbuild/amdModulesWithOut/prepend-reports-deprecation-error.js:461]: It is not encoding file names.. thats the readable baseline :) where first column is fileId encoded in tsbuildInfo and fileNames it refers to as second member of tuple

Comment (sheetalkamat) [reply tests/baselines/reference/tsbuild/amdModulesWithOut/prepend-reports-deprecation-error.js:461]: Look for root in the tsbuildInfo: https://github.com/microsoft/TypeScript/pull/52417/commits/c6d40cb7d62342ada63c4776bec2aad4a997f6b5#diff-6cf9dfab853b1fefb0fdd6192a254278dcd5f9f9e0cdd130503a7b756622473cR68

Comment (jakebailey) [reply tests/baselines/reference/tsbuild/amdModulesWithOut/prepend-reports-deprecation-error.js:461]: Oh! I thought it was just the same thing pretty printed.

**Review by jakebailey — APPROVED** [review]

**Pull request merged**